atc/exec/task_step.go

```diff
@@ -176,6 +173,26 @@ func (step *TaskStep) Run(ctx context.Context, state RunState) error {
 		config.Limits.Memory = step.defaultLimits.Memory
 	}
 
+	events := make(chan runtime.Event, 1)
+	go func(logger lager.Logger, config atc.TaskConfig, events chan runtime.Event, delegate TaskDelegate) {
+		for {
+			ev := <-events
+			switch {
+			case ev.EventType == runtime.InitializingEvent:
+				step.delegate.Initializing(logger, config)
+
+			case ev.EventType == runtime.StartingEvent:
+				step.delegate.Starting(logger, config)
+
+			case ev.EventType == runtime.FinishedEvent:
+				step.delegate.Finished(logger, ExitStatus(ev.ExitStatus))
```

Comment:
I think this `goroutine` is going to leak. You can either `return` after the finish event (which may or may not get called?) or you can [listen for events on a separate channel which gets closed when the surrounding method terminates](https://github.com/concourse/concourse/blob/master/atc/engine/engine.go#L175-L180)

Reply:
@jwinters The channel is closed on line 235 below, would that not go through the `default` branch? I can put line 235 in a `defer` to ensure that it gets closed regardless of errors.

Reply:
Changed it to use [range and close](https://tour.golang.org/concurrency/4) instead